File: python/clase-16/final/main.py
"""API REST con FastAPI, API key de app y variables listas para Railway."""

import os
from anthropic import Anthropic
from fastapi import FastAPI, Header, HTTPException
from pydantic import BaseModel

MODEL = "claude-sonnet-4-6"
app = FastAPI(title="Curso Claude API")
APP_API_KEY = os.getenv("APP_API_KEY")


class ChatRequest(BaseModel):
    message: str


class ChatResponse(BaseModel):
    reply: str


@app.get("/health")
def health() -> dict[str, str]:
    return {"status": "ok"}


@app.post("/chat", response_model=ChatResponse)
def chat(payload: ChatRequest, x_api_key: str = Header(default="")) -> ChatResponse:
    if not APP_API_KEY:
        raise HTTPException(status_code=500, detail="APP_API_KEY no configurada.")
    if x_api_key != APP_API_KEY:
        raise HTTPException(status_code=401, detail="Unauthorized")

    api_key = os.getenv("ANTHROPIC_API_KEY")
    if not api_key:
        raise HTTPException(status_code=500, detail="ANTHROPIC_API_KEY no configurada.")

    client = Anthropic(api_key=api_key)
    response = client.messages.create(
        model=MODEL,
        max_tokens=600,
        messages=[{"role": "user", "content": payload.message}],
    )
    reply = "".join(block.text for block in response.content if block.type == "text")
    return ChatResponse(reply=reply)

File: typescript/clase-16/final/main.ts
import Anthropic from "@anthropic-ai/sdk";
import Fastify from "fastify";
import { z } from "zod";

export {};

const MODEL = "claude-sonnet-4-6";
const APP_API_KEY = process.env.APP_API_KEY;
const ChatRequest = z.object({ message: z.string().min(1) });
const server = Fastify({ logger: true });

server.get("/health", async () => ({ status: "ok" }));

server.post("/chat", async (request, reply) => {
  if (!APP_API_KEY) return reply.code(500).send({ error: "APP_API_KEY no configurada." });
  if (request.headers["x-api-key"] !== APP_API_KEY) {
    return reply.code(401).send({ error: "Unauthorized" });
  }

  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) return reply.code(500).send({ error: "ANTHROPIC_API_KEY no configurada." });

  const payload = ChatRequest.parse(request.body);
  const client = new Anthropic({ apiKey });
  const response = await client.messages.create({
    model: MODEL,
    max_tokens: 600,
    messages: [{ role: "user", content: payload.message }],
  });
  const replyText = response.content.filter((block) => block.type === "text").map((block) => block.text).join("");
  return { reply: replyText };
});

await server.listen({ port: Number(process.env.PORT ?? 3000), host: "0.0.0.0" });
